staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -70,37 +72,33 @@ type CELSchemaContext struct {
 	TotalCost *TotalCost
 }
 
-// HasValidationRules returns true if the current schema node contains x-kubernetes-validations rules.
-func (c *CELSchemaContext) HasValidationRules() bool {
-	return len(c.jsonSchema.XValidations) > 0
-}
-
 // TypeInfo returns the CELTypeInfo for this CELSchemaContext node.  Returns nil if current level or above does not have
-// x-kubernetes-validations rules.
-func (c *CELSchemaContext) TypeInfo() *CELTypeInfo {
-	if !c.withinValidationRuleScope {
-		return nil
+// x-kubernetes-validations rules. The returned type info is shared and should not be modified by the caller.
+func (c *CELSchemaContext) TypeInfo() (*CELTypeInfo, error) {
+	if c == nil || !c.withinValidationRuleScope {
+		return nil, nil
 	}
-	if c.typeInfo != nil {
-		return c.typeInfo
+	if c.typeInfo != nil || c.typeInfoErr != nil {
+		return c.typeInfo, c.typeInfoErr
 	}
 
 	// If able to get the type info from the parent's type info, prefer this approach
 	// since it is more efficient.
 	if c.parent != nil {
-		parentTypeInfo := c.parent.TypeInfo()
+		parentTypeInfo, parentErr := c.parent.TypeInfo()
+		if parentErr != nil {
+			c.typeInfoErr = parentErr
+			return nil, parentErr
+		}
 		if parentTypeInfo != nil && c.typeInfoAccessor != nil {
 			c.typeInfo = c.typeInfoAccessor.accessTypeInfo(parentTypeInfo)
```

Comment:
Some `accessTypeInfo` implementations can return nil (like `propertyTypeInfoAccessor#accessTypeInfo`)... don't we need to check for that and fall through to the expensive `c.typeInfo, c.typeInfoErr = c.converter` path in that case?

Do we have a test case for x-kubernetes-validations rules inside a property with an unescapable name? I'm not seeing how that would work with this code as-is (since it looks like it would try to get the typeInfo from the parent, fail because the property name is not escapable, then not fall through to the path below to get the schema for the level with the x-kubernetes-validations rules)

Reply:
Thank you!  It's a bug and there is a test gap.

I've fleshed out the tests to have all combinations of:

- {unescaped property name ("f"), escaped property name ("f/2"), unescapable property name ("f@2"), items, additionalProperties}
- {parent has schema, parent does not have schema}
- {CRD validation phase, CR validation phase}

It's quite a few more tests, but it catches this bug and I'll verify that the coverage through the new code should now be more comprehensive.